Refactors GF to truly use the singleton pattern

Drake-Eidukas opened the pull request:
Currently, this is the state of `GF`:

```
g1 = GF(1)
g2 = GF(1)
assert g1 is g2 # fails assertion, they're different objects

g3 = GF.get(1)
g4 = GF.get(1)
assert g3 is g4 # passes assertion, intended functionality
```

This PR changes the way that we create `GF` objects, such that both assertions will pass and you can use either method of construction `GF` objects.

For more details, see #226

Drake-Eidukas commented:
Also, I couldn't get VSCode to save the file without changing line 414... but I'm pretty sure that this is the same functionality with a different format.

smkuls commented on honeybadgermpc/field.py:
```diff
@@ -109,15 +118,7 @@ def __reduce__(self):
 
     @staticmethod
     def get(modulus):
```
* This is redundant now. We should get rid of this. 
* I also want you to run `honeybadgermpc/ipc.py` using the `launch.sh` script. I remember we faced some issue when pickling and unpickling. As of now that's the only way to exercise that code (there are no automated tests for that).

Drake-Eidukas replied:
We should consider adding running that script in travis, but it seems to run just fine

smkuls replied:
Yea, we were discussing running the script using a python wrapper and then making sure that things passed.

amiller reviewed (approved):
Nice fix

smkuls commented on honeybadgermpc/field.py:
```diff
@@ -98,7 +98,16 @@ def __int__(self):
 
+    def __new__(cls, modulus):
+        if modulus not in GF._field_cache:
+            GF._field_cache[modulus] = super(GF, cls).__new__(cls)
```
Can we use the `setdefault` method of `dict`?

smkuls commented on honeybadgermpc/field.py:
```diff
@@ -410,7 +411,7 @@ def __init__(self, value):
 
         # Bit extraction. We pretend that the number is *very* big.
-        bit = lambda self, index: 1     # noqa  XXX for the time being
+        def bit(self, index): return 1     # noqa  XXX for the time being
```
Any reason why you changed this?

smkuls replied:
Just saw your comment on the PR, closing this.

Drake-Eidukas committed:
Refactor GF to follow the multiton pattern, removes GF.get()

This commit does 3 things:
- It implements the __new__ method so that you can simply call the constructor
  of GF instead of requiring the use of GF.get(). This will be less error prone.
- It removes the GF.get() method, and refactors existing code to use the constructor
- Adds extra documentation to GF, as well as a new unit test to test the multiton implementation

Drake-Eidukas commented:
Should be good to go now-- implemented samarth's suggestions

smkuls reviewed (approved):
LGTM!
